### PR TITLE
[WIP] Fix py-astropy header for wcslib

### PR DESCRIPTION
**Pull request by LDAmorim**
Hi all,

This small PR tries to address issue #14679 .
The first part of the error came from the file `wcslib.h` being located in `<wcslib-prefix>/include/wcs.h` instead of `<wcslib-prefix>/include/wcslib/wcs.h`.

I tried to use the `find_headers()` function and mimic what was in done for the `OpenMPI` package.py file.
But I am unable to make it work.

I also can't seem to use `debug_print()` or `print()` to see in the shell if my changes are actually being read.

Do you have any suggestions?
Thanks,
Diana

### CODE REVIEW

**Comment by LDAmorim** [comment]
Not sure if I need to add something like:
```python
hdrs.directories = os.path.dirname(hdrs[0])
```

**Comment by scheibelp** [comment]
The approach you are currently taking looks good: you can tell `py-astropy` where to find `wcslib` instead of helping `wcslib` tell dependents where it is.

Regarding the initial approach using `headers`:

`petsc` has an example of `find_headers`, I think the issue with the earlier approach is that `find_headers` expects the name of the header file, which in this case is `wcs.h` but it was called with `wcslib`.

I think the following might work for `wcslib`:

```
def headers(self):
  return find_headers('wcs', self.prefix.include, recursive=True) \
    or None  # return None to indicate failure
```

you shouldn't have to set `.directories` in that case

File: var/spack/repos/builtin/packages/py-astropy/package.py [hunk 1]
```diff
@@ -72,3 +71,9 @@ def build_args(self, spec, prefix):
             args.extend(['-j', str(make_jobs)])
 
         return args
+
+        if '+wcslib' in spec:
```
Comment (scheibelp) [inline var/spack/repos/builtin/packages/py-astropy/package.py]: The dependency on `wcslib` is not conditional, so you can remove this `if` (i.e. always do the below update).

Comment (LDAmorim) [reply var/spack/repos/builtin/packages/py-astropy/package.py]: Oh yes, that makes sense, sorry!

**Comment by LDAmorim** [comment]
> I think the following might work for wcslib:
> 
> def headers(self):
>   return find_headers('wcs', self.prefix.include, recursive=True) \
>     or None  # return None to indicate failure
> you shouldn't have to set .directories in that case

I will try this thank you!
It does make sense since we are looking for `wcs.h` that we look for `wcs` and not `wcslib`.
But, I think that there is another spack package called `wcs` (right above `wcslib` in https://spack.readthedocs.io/en/latest/package_list.html).
If someone has both packages `wcs` and `wcslib` installed via spack, will those lines differentiate between them?

**Comment by LDAmorim** [comment]
It seems that the issue persists even when I added:

```python
def headers(self):
        return find_headers('wcs', self.prefix.include, recursive=True) \
          or None  # return None to indicate failure
```

```bash
  >> 1245    astropy/wcs/include/astropy_wcs/util.h:21:10: fatal error: 'wcs.h' file not found
```

**Comment by scheibelp** [comment]
There's a couple possibilities for what's going wrong:

* I think `-use-system-wcslib` might be useful - I think that is telling astropy to use an external wcslib vs. vendoring it itself - can you try adding that back?
* I'm wondering if we may have to symlink the header directory after all: I'm looking at the `py-astropy` setup for `wcslib` and it might have hardcoded an expected path assuming config files are in `include/` (vs. `include/wcslib/`): `wcsconfig_h_path = join(WCSROOT, 'include', 'wcsconfig.h')`

**Comment by LDAmorim** [comment]
I tested your first point, and the error is still the same.
I don't know how to symlink, but I can try to understand it.
The packaged `Casacore` also uses `wcslib`:
https://github.com/spack/spack/blob/4c7e52adaa37a380c7873eb914b4fcbed8e97fa6/var/spack/repos/builtin/packages/casacore/package.py#L37
So I think it is still best to make the changes to `py-astropy` instead of `wcslib`, just not to compromise packages that might work well right now.
What do you think?

**Comment by scheibelp** [comment]
> So I think it is still best to make the changes to py-astropy instead of wcslib, just not to compromise packages that might work well right now.
> What do you think?

That sounds good, although if the `def headers` approach had worked I would consider it "safe".

> I don't know how to symlink, but I can try to understand it.

The idea would be to create a symlink from `<wcslib-prefix>/include/` to `<wcslib-prefix>/include/wcslib/`. This could be accomplished by adding the following to the `wcslib` package.py:

```
@run_after('install')
def symlink_include(self):
  os.symlink(
    os.path.join(self.prefix, 'include', 'wcslib'),
    os.path.join(self.prefix, 'include')
  )
```

**Comment by adamjstewart** [comment]
I'm confused, I think the `def headers` needs to be added to the `wsclib` package, not the `py-astropy` package. It is then called in the `py-astropy`. Let me know if you need help debugging.

**Comment by scheibelp** [comment]
> I'm confused, I think the def headers needs to be added to the wsclib package, not the py-astropy package. It is then called in the py-astropy. Let me know if you need help debugging.

Yes, I completely missed that. If this will work it will have to be in the `wcslib` package.

**Comment by LDAmorim** [comment]
Hi, sorry for my delay in this.
And thank you for the great feedback.
I can try to test @scheibelp 's lines above asap.
And also add the `def headers` part to the `wcslib` package instead of `py-astropy` one as @adamjstewart suggested.

**Comment by LDAmorim** [comment]
Hi, I still have to test the latest changes, but I am having a new error while installing berkeley-db (that I am trying to figure out - probably am doing something wrong or forgetting a step).

I will update here as soon as I can test the installation of `py-astropy`.
Thanks!

File: var/spack/repos/builtin/packages/wcslib/package.py [hunk 2]
```diff
@@ -47,8 +47,19 @@ def configure_args(self):
...
         # The shared library is not installed correctly on Darwin; fix this
         if self.spec.satisfies('platform=darwin'):
             fix_darwin_install_name(self.prefix.lib)
+
+    @run_after('install') #Is this duplicate necessary?
+    def symlink_include(self):
+        os.symlink(
+            os.path.join(self.prefix, 'include', 'wcslib'),
+            os.path.join(self.prefix, 'include')
+        )
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/wcslib/package.py:65]: I don't love this solution. Packages should always store their headers in a subdirectory, and packages that depend on them should respect this. By putting everything in `self.prefix.include`, you're risking conflicting files when multiple packages are merged to the same directory.

I'm going to investigate this further. Maybe Astropy 4.0 fixes this problem. Or a different version of wcslib is required. I'm also seeing issues with `berkeley-db` though, at least on macOS.

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/wcslib/package.py:65]: I think the right solution is to add a build dependency on pkg-config. I'm also adding wcslib 7.3 and astropy 4.0, so I'll make new PRs for these.

**Comment by scheibelp** [comment]
> I think the right solution is to add a build dependency on pkg-config.

That didn't occur to me that it might optionally use `pkg-config`. For dependents which do not use `pkgconfig` to find resources, this PR may be required. It will also ensure that the `-I` entries inserted by the Spack compiler wrappers are correct. I agree that when a dependent uses `pkgconfig` it would be a preferable solution to depending on `headers`.

> I don't love this solution [introducing the symlink]. Packages should always store their headers in a subdirectory, and packages that depend on them should respect this

I think the `headers` definition may have been sufficient (i.e. the symlink may not be required now that `headers` is in the right place). I think I see what you mean that the symlink should be avoided (avoid conflicting header files).

I'm inclined to 

* Merge #17838
* Leave this open

@LDAmorim feel free to close this if you are short on time to work on it, since #17838 fixes the immediate issue. To test the edits here with `py-astropy` you would need to remove the `pkg-config` dependency that #17838 introduces there).

**Comment by adamjstewart** [comment]
> For dependents which do not use pkgconfig to find resources, this PR may be required. It will also ensure that the -I entries inserted by the Spack compiler wrappers are correct.

The problem here is, what is correct? I'm not a C/C++ person, but my understanding is that packages tend to put their headers in subdirectories like `prefix/include/wcslib/wcs.h` and that packages that include these headers should use `#include <wcslib/wcs.h>`, but some packages like astropy choose to use `#include <wcs.h>`, so it isn't clear to me what the correct solution is.

**Comment by LDAmorim** [comment]
Hi,
Sorry I wasn't able to reply before (I have been at a training these past weeks).

I just got my macOS build up to date to test this fix following @scheibelp 's suggestion:

> @LDAmorim feel free to close this if you are short on time to work on it, since #17838 fixes the immediate issue. To test the edits here with py-astropy you would need to remove the pkg-config dependency that #17838 introduces there).

I commented out the `pkg-config` dependency and tried to install `wcslib` (after I would try `py-astropy`) with the changes done here and the system described below.

But other packages install `pkgconf` even before `wcslib`. I am not sure if that is the reason, but I can't succeed in the installation phase:

```console
 $ spack install wcslib
==> Warning: apple-clang@11.0.3 cannot build optimized binaries for "skylake". Using best target possible: "x86_64"
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/libsigsegv-2.12-cd2u6jt6oagbnhsgqal54myovfkwpri6
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/berkeley-db-18.1.40-3f4556kvgg2ihsahtuy4ta27i65hnikf
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/pkgconf-1.7.3-52kfpvjixcntquvt67sltjuajdhq73od
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/libiconv-1.16-2owd6xnxysuqdfuuu4mgemxmdhgto4px
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/xz-5.2.5-nfsvq5xchnmyvqc333p644ouuwnk7awe
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/zlib-1.2.11-7erteezzqncanvj4cr4lesgcawuztojo
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/m4-1.4.18-anoponiwofd7dqkxmcqy6iq6ildhe5s2
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/ncurses-6.2-wr2lahob7rnaawch75w2p5rce4qk3qul
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/tar-1.32-hn6spwek3t2u4da6awnmee6q7xk26qrj
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/diffutils-3.7-qx2vqhptr5rdek26dl5a27hbnyxsouff
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/libxml2-2.9.10-cvsddxrqg4jwdgk2xopykzztgwwypl3f
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/libtool-2.4.6-yrvttuduwubr6aebtif4ezrllojrfi7i
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/readline-8.0-mtiikfoyt34p7bvxqz3talk7xv5azul4
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/bzip2-1.0.8-uyeu5febth5zmixtjboghc7lz5533zms
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gdbm-1.18.1-aaqin73b6t3wbxht7pdtw3ozteqmecd2
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gettext-0.20.2-syawspfobbr27xxhxwhbrzlgo5iun5sw
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/perl-5.30.3-7aafes7kqahuojjmxcqtwdezuqh74ya4
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/help2man-1.47.11-3gttfzbt5t2t2mr2kdgtlx6zklgbealm
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/texinfo-6.5-bpmdlwfjdbg3dynlm5hjtnhtohpcaqyy
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/autoconf-2.69-caqezswjz5nwgqaxltqjnk25qjsuszfo
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/bison-3.6.4-kclbqghyunp7ra5yflsjrnsoihd3bykp
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gmake-4.2.1-tmmxdeupufcqa2f6xskefhwyqdcmr6a4
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/automake-1.16.2-2ug6frtywzpnmtbh2uji4pllbcgbbeoy
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/findutils-4.6.0-2swrp2onomc2norjkbmshw5xajlxrerw
[+] /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/flex-2.6.4-nuyisn7iedyjqh6gjcvbxgd4ecn7nzrz
==> Installing wcslib
==> No binary for wcslib found: installing from source
==> Warning: microarchitecture specific optimizations are not supported yet on mixed compiler toolchains [check apple-clang@11.0.3 for further details]
==> wcslib: Executing phase: 'autoreconf'
==> wcslib: Executing phase: 'configure'
==> wcslib: Executing phase: 'build'

==> wcslib: Executing phase: 'install'
==> Error: FileExistsError: [Errno 17] File exists: '/Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/wcslib-6.4-6nbsz25ig6ndimyz45aa5i2aoqbqeejd/include/wcslib' -> '/Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/wcslib-6.4-6nbsz25ig6ndimyz45aa5i2aoqbqeejd/include'

/Users/LDianaAmorim/Documents/opt/spack/var/spack/repos/builtin/packages/wcslib/package.py:64, in symlink_include:
         61    def symlink_include(self):
         62        os.symlink(
         63            os.path.join(self.prefix, 'include', 'wcslib'),
  >>     64            os.path.join(self.prefix, 'include')
         65        )

```

It seems like although these lines are for after the installation process, they are called before the `wcslib` folder exists even.
Do you have any suggestions to fix this?

When I go back to the `develop` branch I don't get any issue, so your fix in #17838 also works, thank your for it!

Regarding your last comment @adamjstewart , would you like me to contact py-astropy and ask them to change the include on their side?

Thanks,
Diana

### My system

```console
$ spack debug report
* **Spack:** 0.15.3-421-c83236d3a
* **Python:** 3.7.3
* **Platform:** darwin-catalina-skylake
```

* clang version: 11.0.3
* gcc (gfortran) version: 10.2.0
* cmake version: 3.18.1



```console
$ cat ~/.spack/darwin/packages.yaml
packages:
  cmake:
    buildable: False
    paths:
      cmake@3.18.1: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/cmake-3.18.1-u4guucbahyuq5pswf55b2ebqslsjub7m
```

```console
$ % cat ~/.spack/darwin/compilers.yaml 
compilers:
- compiler:
    spec: apple-clang@11.0.3
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/gfortran
      fc: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/gfortran
    flags: {}
    operating_system: catalina
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
- compiler:
    spec: gcc@10.2.0
    paths:
      cc: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/gcc
      cxx: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/g++
      f77: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/gfortran
      fc: /Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/gcc-10.2.0-phyqkhktab6nxd2noi3kklzjxsnku5le/bin/gfortran
    flags: {}
    operating_system: catalina
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

**Comment by adamjstewart** [comment]
> ==> Error: FileExistsError: [Errno 17] File exists: '/Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/wcslib-6.4-6nbsz25ig6ndimyz45aa5i2aoqbqeejd/include/wcslib' -> '/Users/LDianaAmorim/Documents/opt/spack/opt/spack/darwin-catalina-x86_64/apple-clang-11.0.3/wcslib-6.4-6nbsz25ig6ndimyz45aa5i2aoqbqeejd/include'

The error is that you're trying to symlink `prefix.include.wcslib` to `prefix.include`, but that file already exists. Try getting rid of the symlinking step, I don't think that's the right idea.

**Comment by LDAmorim** [comment]
Sorry, yes I was confusing that part.
Now I understand.

I checked without the `symlink` part, and got:
```console
==> Installing py-astropy
==> No binary for py-astropy found: installing from source
==> Warning: microarchitecture specific optimizations are not supported yet on mixed compiler toolchains [check apple-clang@11.0.3 for further details]
==> Error: AttributeError: 'function' object has no attribute 'directories'

/Users/LDianaAmorim/Documents/opt/spack/lib/spack/spack/build_environment.py:862, in child_process:
        859            tb_string = traceback.format_exc()
        860
        861            # build up some context from the offending package so we can
  >>    862            # show that, too.
        863            package_context = get_package_context(tb)
        864
        865            build_log = None

```

**Comment by LDAmorim** [comment]
Hi all,
I am so sorry that this PR is still open.
I had to move and now have a Windows/Ubunto machine so I won`t be able to test this on a macOS at anytime soon.
I think that since no one is needing it anymore, maybe I can close this PR.
What do you think?
Thank you,
Diana

**Comment by adamjstewart** [comment]
I think it's safe to close this one. #14679 was fixed by #17838, so this PR is no longer needed.

**Pull request closed without merging**